sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/aggregator/DoubleExponentialHistogramBuckets.java

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.internal.aggregator;
+
+import io.opentelemetry.sdk.metrics.data.ExponentialHistogramBuckets;
+import io.opentelemetry.sdk.metrics.internal.state.ExponentialCounter;
+import io.opentelemetry.sdk.metrics.internal.state.MapCounter;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Logger;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * This class handles the operations for recording, scaling, and exposing data related to the
+ * exponential histogram.
+ *
+ * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
+ * at any time.
+ */
+final class DoubleExponentialHistogramBuckets implements ExponentialHistogramBuckets {
+
+  public static final int MAX_SCALE = 20;
+
+  private static final int MAX_BUCKETS = MapCounter.MAX_SIZE;
+  private static final Logger logger =
+      Logger.getLogger(DoubleExponentialHistogramBuckets.class.getName());
+
+  private ExponentialCounter counts;
+  private BucketMapper bucketMapper;
+  private int scale;
+
+  DoubleExponentialHistogramBuckets() {
+    this.counts = new MapCounter();
+    this.bucketMapper = new LogarithmMapper(MAX_SCALE);
+    this.scale = MAX_SCALE;
+  }
+
+  // For copying
+  DoubleExponentialHistogramBuckets(DoubleExponentialHistogramBuckets buckets) {
+    this.counts = new MapCounter(buckets.counts); // copy counts
+    this.bucketMapper = new LogarithmMapper(buckets.scale);
+    this.scale = buckets.scale;
+  }
+
+  public boolean record(double value) {
+    int index = bucketMapper.valueToIndex(Math.abs(value));
+    return this.counts.increment(index, 1);
+  }
+
+  @Override
+  public int getOffset() {
+    return counts.getIndexStart();
+  }
+
+  @Nonnull
+  @Override
+  public List<Long> getBucketCounts() {
+    if (counts.isEmpty()) {
+      return Collections.emptyList();
+    }
+    int length = counts.getIndexEnd() - counts.getIndexStart() + 1;
+    Long[] countsArr = new Long[length];
+    for (int i = 0; i < length; i++) {
+      countsArr[i] = counts.get(i + counts.getIndexStart());
+    }
+    return Arrays.asList(countsArr);
+  }
+
+  @Override
+  public long getTotalCount() {
+    long totalCount = 0;
+    for (int i = counts.getIndexStart(); i <= counts.getIndexEnd(); i++) {
+      totalCount += counts.get(i);
+    }
+    return totalCount;
+  }
+
+  void downscale(int by) {
+    if (by == 0) {
+      return;
+    } else if (by < 0) {
+      logger.warning(
+          "downScale() expects non-negative integer but was given"
```

Comment:
Is this a programming bug in our codebase? Then we can throw IllegalStateException instead. Otherwise probably can add a test case

Reply:
I believe this should be IllegalStateException.

Reply:
Done!